File: src/playwright/controller.ts
import path from 'node:path';
import { BrowserPlatformError } from '../core/errors.js';
import type { SessionActionPayload } from '../daemon/types.js';
import { TraceWriter } from '../traces/writer.js';
import {
  isDebugEnabled,
  captureDebugStep,
  captureDebugStepJson,
  appendDebugLog
} from '../debug/capture.js';
import {
  BrowserContextPool,
  BrowserSession,
  type AdoptedBrowserSession,
  type BrowserSessionOpenResult,
  type BrowserSessionSnapshotResult,
  type PageStateSummary
} from './browser-session.js';
import { buildActionResult, runStep } from '../runtime/run-step.js';
import type { SessionBackend } from '../daemon/types.js';
import type { LoadedSitePack } from '../packs/loader.js';

export class PlaywrightController {
  private readonly sessions = new Map<string, BrowserSession>();
  private readonly sessionOperationTails = new Map<string, Promise<void>>();
  private readonly lastObservations = new Map<
    string,
    { state: PageStateSummary; observedAtMs: number }
  >();
  private readonly traceWriter: TraceWriter;
  private readonly contextPool = new BrowserContextPool();

  constructor(private readonly rootDir: string) {
    this.traceWriter = new TraceWriter(
      path.join(this.rootDir, 'artifacts', 'traces')
    );
  }

  async openSession(
    sessionId: string,
    url: string,
    options?: {
      storageStatePath?: string;
      backend?: SessionBackend;
    }
  ): Promise<BrowserSessionOpenResult> {
    const session = new BrowserSession({
      sessionId,
      snapshotRootDir: path.join(this.rootDir, 'artifacts', 'snapshots'),
      storageStatePath: options?.storageStatePath,
      backend: options?.backend,
      contextPool: options?.storageStatePath ? this.contextPool : undefined
    });

    const opened = await session.open(url);
    session.markUsed();
    this.sessions.set(sessionId, session);
    await this.debugCapture(sessionId, 'open', {
      sessionId,
      url: opened.url,
      title: opened.title
    });
    return opened;
  }

  async observeSession(sessionId: string): Promise<PageStateSummary> {
    return this.runExclusive(sessionId, 'observe', async () => {
      const session = this.requireSession(sessionId);
      session.markUsed();
      const result = await session.observe();
      this.lastObservations.set(sessionId, {
        state: result,
        observedAtMs: Date.now()
      });
      await this.debugCapture(sessionId, 'observe', {
        sessionId,
        url: result.url,
        title: result.title
      });
      return result;
    });
  }

  async adoptSession(
    sessionId: string,
    adopted: AdoptedBrowserSession,
    options?: {
      storageStatePath?: string;
      backend?: SessionBackend;
    }
  ): Promise<BrowserSessionOpenResult> {
    return this.runExclusive(sessionId, 'adopt', async () => {
      await this.closeSessionUnlocked(sessionId);

      const session = new BrowserSession({
        sessionId,
        snapshotRootDir: path.join(this.rootDir, 'artifacts', 'snapshots'),
        storageStatePath: options?.storageStatePath,
        backend: options?.backend
      });

      try {
        session.adoptExisting(adopted);
        session.markUsed();
        await session.persistStorageState({ force: true });
        const page = session.page();
        this.sessions.set(sessionId, session);

        return {
          url: page.url(),
          title: await page.title()
        };
      } catch (error) {
        await session.close().catch(() => undefined);
        throw error;
      }
    });
  }

  async writeTrace(sessionId: string, stepType: string, payload: unknown) {
    return this.traceWriter.writeStep(sessionId, stepType, payload);
  }

  async flushTraces(): Promise<void> {
    await this.traceWriter.flush();
  }

  async actInSession(
    sessionId: string,
    payload: SessionActionPayload,
    options: { sitePack?: LoadedSitePack | null } = {}
  ) {
    return this.runExclusive(sessionId, 'act', async () => {
      const session = this.requireSession(sessionId);
      session.markUsed();
      const cached = this.lastObservations.get(sessionId);
      const beforeFromCache =
        cached && Date.now() - cached.observedAtMs <= 1_000
          ? cached.state
          : undefined;
      const { before, after, observations } = await runStep(session, payload, {
        before: beforeFromCache,
        sitePack: options.sitePack
      });
      this.lastObservations.set(sessionId, {
        state: after,
        observedAtMs: Date.now()
      });
      const result = buildActionResult(payload, before, after, observations);
      await this.debugCapture(sessionId, `act-${payload.action}`, {
        sessionId,
        action: result.action,
        target: result.target,
        input: result.input,
        before: { url: before.url, title: before.title },
        after: { url: after.url, title: after.title },
        changes: result.changes
      });
      return result;
    });
  }

  async snapshotSession(
    sessionId: string
  ): Promise<BrowserSessionSnapshotResult> {
    return this.runExclusive(sessionId, 'snapshot', async () => {
      const session = this.requireSession(sessionId);
      session.markUsed();
      const result = await session.snapshot();
      this.lastObservations.set(sessionId, {
        state: result.state,
        observedAtMs: Date.now()
      });
      if (isDebugEnabled()) {
        await captureDebugStepJson(this.rootDir, sessionId, 'snapshot', {
          sessionId,
          screenshotPath: result.screenshotPath,
          htmlPath: result.htmlPath,
          url: result.state.url,
          title: result.state.title
        });
      }
      return result;
    });
  }

  async closeSession(sessionId: string): Promise<void> {
    await this.runExclusive(sessionId, 'close', async () => {
      await this.closeSessionUnlocked(sessionId);
    });
  }

  async persistSessionStorageState(
    sessionId: string,
    options: { force?: boolean } = {}
  ): Promise<void> {
    await this.runExclusive(sessionId, 'persist-storage-state', async () => {
      const session = this.requireSession(sessionId);
      session.markUsed();
      await session.persistStorageState(options);
    });
  }

  private async closeSessionUnlocked(sessionId: string): Promise<void> {
    const session = this.sessions.get(sessionId);
    if (!session) {
      return;
    }

    this.sessions.delete(sessionId);
    this.lastObservations.delete(sessionId);
    await session.close();
  }

  private async runExclusive<T>(
    sessionId: string,
    opName: string,
    fn: () => Promise<T>
  ): Promise<T> {
    const queuedAtMs = Date.now();
    const queuedAt = new Date(queuedAtMs).toISOString();
    const previousTail =
      this.sessionOperationTails.get(sessionId) ?? Promise.resolve();
    const operationPromise = previousTail
      .catch(() => undefined)
      .then(async () => {
        const startedAtMs = Date.now();
        const startedAt = new Date(startedAtMs).toISOString();
        let status: 'ok' | 'error' = 'ok';

        try {
          return await fn();
        } catch (error) {
          status = 'error';
          throw error;
        } finally {
          const finishedAtMs = Date.now();
          if (isDebugEnabled()) {
            void appendDebugLog(this.rootDir, {
              source: 'browser',
              event: 'session-operation',
              sessionId,
              opName,
              status,
              queuedAt,
              startedAt,
              finishedAt: new Date(finishedAtMs).toISOString(),
              waitedMs: startedAtMs - queuedAtMs,
              runMs: finishedAtMs - startedAtMs
            });
          }
        }
      });

    const tail = operationPromise.then(
      () => undefined,
      () => undefined
    );
    this.sessionOperationTails.set(sessionId, tail);

    try {
      return await operationPromise;
    } finally {
      if (this.sessionOperationTails.get(sessionId) === tail) {
        this.sessionOperationTails.delete(sessionId);
      }
    }
  }

  hasSession(sessionId: string): boolean {
    return this.sessions.has(sessionId);
  }

  async closeAll(): Promise<void> {
    const sessionIds = [...this.sessions.keys()];
    await Promise.all(
      sessionIds.map((sessionId) => this.closeSession(sessionId))
    );
    await this.contextPool.closeAll();
  }

  getSessionPage(sessionId: string): import('playwright').Page | null {
    return this.sessions.get(sessionId)?.page() ?? null;
  }

  private async debugCapture(
    sessionId: string,
    stepName: string,
    meta: unknown
  ): Promise<void> {
    if (!isDebugEnabled()) return;
    const startMs = Date.now();
    const page = this.sessions.get(sessionId)?.page();
    if (!page) return;
    await captureDebugStep(page, this.rootDir, sessionId, stepName, meta);
    await appendDebugLog(this.rootDir, {
      source: 'browser',
      event: stepName,
      sessionId,
      durationMs: Date.now() - startMs,
      ...(meta !== null && typeof meta === 'object'
        ? (meta as Record<string, unknown>)
        : { meta })
    });
  }

  private requireSession(sessionId: string): BrowserSession {
    const session = this.sessions.get(sessionId);
    if (!session) {
      throw new BrowserPlatformError('Session not found', {
        code: 'SESSION_NOT_FOUND'
      });
    }

    return session;
  }
}

File: browser_platform/playwright/controller.py
from __future__ import annotations
import asyncio
import time
from pathlib import Path

from ..core.errors import BrowserPlatformError
from ..traces.writer import TraceWriter
from ..debug.capture import (
    is_debug_enabled,
    capture_debug_step,
    capture_debug_step_json,
    append_debug_log,
)
from .browser_session import BrowserContextPool, BrowserSession


class PlaywrightController:
    def __init__(self, root_dir: str) -> None:
        self._root_dir = root_dir
        self._sessions: dict[str, BrowserSession] = {}
        self._session_locks: dict[str, asyncio.Lock] = {}
        self._trace_writer = TraceWriter(
            str(Path(root_dir) / 'artifacts' / 'traces')
        )
        self._context_pool = BrowserContextPool()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    async def open_session(
        self,
        session_id: str,
        url: str,
        *,
        storage_state_path: str | None = None,
        backend: str | None = None,
    ) -> dict:
        session = BrowserSession(
            session_id=session_id,
            snapshot_root_dir=str(Path(self._root_dir) / 'artifacts' / 'snapshots'),
            storage_state_path=storage_state_path,
            backend=backend,
            context_pool=self._context_pool if storage_state_path else None,
        )
        opened = await session.open(url)
        session.mark_used()
        self._sessions[session_id] = session
        await self._debug_capture(session_id, 'open', {
            'sessionId': session_id,
            'url': opened['url'],
            'title': opened['title'],
        })
        return opened

    async def observe_session(self, session_id: str) -> dict:
        async def _op():
            session = self._require_session(session_id)
            session.mark_used()
            result = await session.observe()
            await self._debug_capture(session_id, 'observe', {
                'sessionId': session_id,
                'url': result['url'],
                'title': result['title'],
            })
            return result

        return await self._run_exclusive(session_id, 'observe', _op)

    async def adopt_session(
        self,
        session_id: str,
        adopted: dict,
        *,
        storage_state_path: str | None = None,
        backend: str | None = None,
    ) -> dict:
        async def _op():
            await self._close_session_unlocked(session_id)

            session = BrowserSession(
                session_id=session_id,
                snapshot_root_dir=str(Path(self._root_dir) / 'artifacts' / 'snapshots'),
                storage_state_path=storage_state_path,
                backend=backend,
            )
            try:
                session.adopt_existing(adopted)
                session.mark_used()
                await session.persist_storage_state()
                page = session.page()
                self._sessions[session_id] = session
                return {
                    'url': page.url,
                    'title': await page.title(),
                }
            except Exception:
                await session.close()
                raise

        return await self._run_exclusive(session_id, 'adopt', _op)

    async def write_trace(self, session_id: str, step_type: str, payload: object) -> dict:
        return self._trace_writer.write_step(session_id, step_type, payload)

    async def act_in_session(self, session_id: str, payload: dict) -> dict:
        async def _op():
            from ..runtime.run_step import run_step, build_action_result
            session = self._require_session(session_id)
            session.mark_used()
            before, after, observations = await run_step(session, payload)
            await session.persist_storage_state()
            result = build_action_result(payload, before, after, observations)
            await self._debug_capture(session_id, f"act-{payload.get('action')}", {
                'sessionId': session_id,
                'action': result.get('action'),
                'target': result.get('target'),
                'input': result.get('input'),
                'before': {'url': before['url'], 'title': before['title']},
                'after': {'url': after['url'], 'title': after['title']},
                'changes': result.get('changes'),
            })
            return result

        return await self._run_exclusive(session_id, 'act', _op)

    async def snapshot_session(self, session_id: str) -> dict:
        async def _op():
            session = self._require_session(session_id)
            session.mark_used()
            result = await session.snapshot()
            if is_debug_enabled():
                await capture_debug_step_json(
                    self._root_dir, session_id, 'snapshot', {
                        'sessionId': session_id,
                        'screenshotPath': result['screenshotPath'],
                        'htmlPath': result['htmlPath'],
                        'url': result['state']['url'],
                        'title': result['state']['title'],
                    }
                )
            return result

        return await self._run_exclusive(session_id, 'snapshot', _op)

    async def close_session(self, session_id: str) -> None:
        async def _op():
            await self._close_session_unlocked(session_id)

        await self._run_exclusive(session_id, 'close', _op)

    def has_session(self, session_id: str) -> bool:
        return session_id in self._sessions

    async def close_all(self) -> None:
        session_ids = list(self._sessions.keys())
        await asyncio.gather(*(self.close_session(sid) for sid in session_ids))
        await self._context_pool.close_all()

    def get_session_page(self, session_id: str):
        session = self._sessions.get(session_id)
        if session is None:
            return None
        try:
            return session.page()
        except Exception:
            return None

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    async def _close_session_unlocked(self, session_id: str) -> None:
        session = self._sessions.pop(session_id, None)
        if session is not None:
            await session.close()

    async def _run_exclusive(self, session_id: str, op_name: str, fn) -> object:
        lock = self._session_locks.setdefault(session_id, asyncio.Lock())
        queued_at_ms = time.time() * 1000
        queued_at = _iso_now()
        async with lock:
            started_at_ms = time.time() * 1000
            started_at = _iso_now()
            status = 'ok'
            try:
                return await fn()
            except Exception:
                status = 'error'
                raise
            finally:
                finished_at_ms = time.time() * 1000
                if is_debug_enabled():
                    asyncio.ensure_future(append_debug_log(self._root_dir, {
                        'source': 'browser',
                        'event': 'session-operation',
                        'sessionId': session_id,
                        'opName': op_name,
                        'status': status,
                        'queuedAt': queued_at,
                        'startedAt': started_at,
                        'finishedAt': _iso_now(),
                        'waitedMs': started_at_ms - queued_at_ms,
                        'runMs': finished_at_ms - started_at_ms,
                    }))

    async def _debug_capture(self, session_id: str, step_name: str, meta: dict) -> None:
        if not is_debug_enabled():
            return
        start_ms = time.time() * 1000
        page = self.get_session_page(session_id)
        if page is None:
            return
        await capture_debug_step(page, self._root_dir, session_id, step_name, meta)
        await append_debug_log(self._root_dir, {
            'source': 'browser',
            'event': step_name,
            'sessionId': session_id,
            'durationMs': time.time() * 1000 - start_ms,
            **meta,
        })

    def _require_session(self, session_id: str) -> BrowserSession:
        session = self._sessions.get(session_id)
        if session is None:
            raise BrowserPlatformError('Session not found', code='SESSION_NOT_FOUND')
        return session


def _iso_now() -> str:
    from datetime import datetime, timezone
    return datetime.now(timezone.utc).isoformat()
